libs/util/httpRequest.js

```diff
@@ -9,154 +9,161 @@
 
 var FetchrError = require('./FetchrError');
 
-function shouldRetry(err, options, attempt) {
+function _shouldRetry(err) {
     if (err.reason === FetchrError.ABORT) {
         return false;
     }
 
-    if (attempt >= options.retry.maxRetries) {
+    if (this._currentAttempt >= this._options.retry.maxRetries) {
         return false;
     }
 
-    if (options.method === 'POST' && !options.retry.retryOnPost) {
+    if (this._options.method === 'POST' && !this._options.retry.retryOnPost) {
         return false;
     }
 
-    return options.retry.statusCodes.indexOf(err.statusCode) !== -1;
+    return this._options.retry.statusCodes.indexOf(err.statusCode) !== -1;
 }
 
-function delayPromise(fn, delay) {
+// _retry is the onReject promise callback that we attach to the
+// _fetch call (ex. _fetch().catch(_retry)). Since _fetch is a promise
+// and since we must be able to retry requests (aka call _fetch
+// function again), we must call _fetch from within _retry. This means
+// that _fetch is a recursive function. Recursive promises are
+// problematic since they can block the main thread for a
+// while. However, since the inner _fetch call is wrapped in a
+// setTimeout we are safe here.
+//
+// The call flow:
+//
+// send -> _fetch -> _retry -> _fetch -> _retry -> end
+function _retry(err) {
+    var self = this;
+    if (!_shouldRetry.call(self, err)) {
+        throw err;
+    }
+
+    // Use exponential backoff and full jitter
+    // strategy published in
+    // https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+    var delay =
+        Math.random() *
+        self._options.retry.interval *
+        Math.pow(2, self._currentAttempt);
+
+    self._controller = new AbortController();
+    self._currentAttempt += 1;
+
     return new Promise(function (resolve, reject) {
         setTimeout(function () {
-            fn().then(resolve, reject);
+            _fetch.call(self).then(resolve, reject);
         }, delay);
     });
 }
 
-function io(options, controller) {
+function _fetch() {
+    var self = this;
```

Comment:
nit: do we still need to use `self`?

Reply:
We need it later in the fetch call itself. Most of the time we could use `this` and only self in the fetch promise. But having both would be annoying so I opted to only use one. I was also very tempted to use arrow functions so we wouldn't need this workaround. I was afraid that you would consider it a breaking change, so I kept the es5 style.